exchanges/stream/websocket.go

```diff
@@ -786,163 +776,134 @@ func (w *Websocket) GetName() string {
+		if s.State() == subscription.InactiveState {
+			if err := s.SetState(subscription.SubscribingState); err != nil {
+				errs = common.AppendError(errs, fmt.Errorf("%w: %s", err, s))
+			}
+		}
+		if err := w.subscriptions.Add(s); err != nil {
+			errs = common.AppendError(errs, err)
+		}
 	}
-	p.State = state
-	return nil
+	return errs
 }
 
-// AddSuccessfulSubscriptions adds subscriptions to the subscription lists that
-// has been successfully subscribed
-func (w *Websocket) AddSuccessfulSubscriptions(channels ...subscription.Subscription) {
-	w.subscriptionMutex.Lock()
-	defer w.subscriptionMutex.Unlock()
+// AddSuccessfulSubscriptions marks subscriptions as subscribed and adds them to the subscription store
+func (w *Websocket) AddSuccessfulSubscriptions(subs ...*subscription.Subscription) error {
+	if w == nil {
+		return common.ErrNilPointer
+	}
 	if w.subscriptions == nil {
-		w.subscriptions = subscriptionMap{}
+		w.subscriptions = subscription.NewStore()
 	}
-	for _, cN := range channels { //nolint:gocritic // See below comment
-		c := cN // cN is an iteration var; Not safe to make a pointer to
-		key := c.EnsureKeyed()
-		c.State = subscription.SubscribedState
-		w.subscriptions[key] = &c
+	var errs error
+	for _, s := range subs {
+		if err := s.SetState(subscription.SubscribedState); err != nil {
+			errs = common.AppendError(errs, fmt.Errorf("%w: %s", err, s))
+		}
+		if err := w.subscriptions.Add(s); err != nil {
+			errs = common.AppendError(errs, err)
+		}
 	}
+	return errs
 }
 
-// RemoveSubscriptions removes subscriptions from the subscription list
-func (w *Websocket) RemoveSubscriptions(channels ...subscription.Subscription) {
-	w.subscriptionMutex.Lock()
-	defer w.subscriptionMutex.Unlock()
-	if w.subscriptions == nil {
-		w.subscriptions = subscriptionMap{}
+// RemoveSubscriptions removes subscriptions from the subscription list and sets the status to Unsubscribed
+func (w *Websocket) RemoveSubscriptions(subs ...*subscription.Subscription) error {
+	if w == nil || w.subscriptions == nil {
+		return common.ErrNilPointer
 	}
-	for i := range channels {
-		key := channels[i].EnsureKeyed()
-		delete(w.subscriptions, key)
+	var errs error
+	for _, s := range subs {
+		if err := s.SetState(subscription.UnsubscribedState); err != nil {
+			errs = common.AppendError(errs, fmt.Errorf("%w: %s", err, s))
+		}
+		if err := w.subscriptions.Remove(s); err != nil {
+			errs = common.AppendError(errs, err)
+		}
 	}
+	return errs
 }
 
-// GetSubscription returns a pointer to a copy of the subscription at the key provided
+// GetSubscription returns a subscription at the key provided
 // returns nil if no subscription is at that key or the key is nil
+// Keys can implement subscription.MatchableKey in order to provide custom matching logic
 func (w *Websocket) GetSubscription(key any) *subscription.Subscription {
-	if key == nil || w == nil || w.subscriptions == nil {
+	if w == nil || w.subscriptions == nil || key == nil {
 		return nil
```

Comment:
suggestion: return error

Reply:
Considered that before (actually, 90% sure you raised this on your first review here).
I think it's acceptable to just return nil.
Something else will catch that it's nil if it's an issue, and realistically we *didn't* find the Sub, so 🤷

Reply:
I have taken John West adage of rejecting fish the appropriate approach. In this edge case your shifting the burden of proof down a callers stack to another websocket method which could potentially do the same thing or; it should be captured before this is called. 🤷 Either way still a suggestion.